[AIFlow] Improve overview.md document. (#97)
flink-extended/ai-flow: docs/content/architecture/overview.md

      
      * Control Edge: It describes the event trigger conditions for controlling Job.
      
    * __Translator__: The role of the Translator is to convert the AI Graph defined by the AIFlow program 
      into a Workflow.
   

> 2. __AIFlow Server__: AIFlow Server provides meta information management and scheduling services.

Reviewer: I think model center should also be metioned.

   * __Meta Service__: Meta service is responsible for managing meta information of AIFlow projects. 
     It includes dataset meta, model meta, workflow meta, artifact meta, metric meta, project meta, etc.
     
   * __Scheduling Service__: Scheduling Service is responsible for processing requests to submit workflow, 
     run workflow, stop workflow, etc.

3. __Notification Server__: Notification Server provides notification service which support the event publishing and subscription.

4. __Scheduler__: Scheduler provides the function of executing workflow.

## Principle

> The figure below shows the principle of AIFlow:

Reviewer 2: ```suggestion
The principle of AIFlow is shown in the figure:
```

> AI Graph: As shown in the figure below, it consists of AI Node, Data Edge and Control Edge . 

Reviewer 2: ```suggestion
AI Graph: As shown in the figure below, the AI Graph is consisted of the AI Node, Data Edge and Control Edge . 
```

> Translator: It converts the AI Graph into the Workflow that the scheduler can execute.

Reviewer 2: ```suggestion
Translator: It translates the AI Graph to the executable workflow for Scheduler.
```

![Alt text](../images/architecture/translator.png)

The working steps of the translator are as follows:
1. A group of AI Nodes with the same job config are combined into an AISubGraph.
2. Call the corresponding JobGenerator according to the job config to convert all AISubGraphs to the corresponding Jobs.
3. Add all Control Edges to the Workflow.

## AIFlow Server

> AIFlow Server provides meta information management and scheduling services.

Reviewer 2: ```suggestion
AIFlow Server supports the Meta Service and Scheduling Service to provide the basic service of the AIFlow.
```

### Meta Service

Meta Service: Provides the CRUD(Create, Read, Update and Delete) service of metadata generated in the workflow of machine learning tasks.

The services provided are shown in the figure below:

![Alt text](../images/architecture/meta.png)

> 1. Dataset: It provides dataset meta-information services, including data address, data format and other information.

Reviewer 2: All `meta-information` could update to `meta`.

   Users can define the program for reading and writing data according to it.
2. Project: It provides project meta-information services, 
   including project name, project description and other information.
   Users can organize the machine learning workflow of different projects based on it.
3. Workflow: It provides workflows meta-information services.
   Users can view the submitted workflow's meta-information according to it.
4. Model: It provides Models meta-information services.
   It contains the description information of the model and the version information of the model.
5. Metric: It provides metric meta-information services, such as model version evaluate metric.
6. Artifact: It provides artifacts meta-information services such as configuration files, jar packages etc.

### Scheduling Service

> Scheduling Service: The Scheduling Service converts the workflow into an object that can be recognized by the scheduler 

Reviewer 2: ```suggestion
Scheduling Service: The Scheduling Service converts the workflow translated into the executable workflow recognized by Scheduler and schedules the workflow execution.
```

and submits it to the scheduler.

> The traditional scheduler can only schedule batch jobs (when upstream job finished then downstream jobs can run) 

Reviewer 2: ```suggestion
The traditional scheduler could only supports the scheduling of the batch jobs(It means that after upstream job finished, downstream jobs could run.)
```

but in the online learning scenario where we have jobs that will never finish, it does not meet the demand. 

> The scheduler for Scheduler Service docking must support event-based scheduling.

Reviewer 2: Add the introduction of the event-based scheduling. The users don't understand what the event-based scheduling is.

The following figure shows the principle of Scheduler Service:

![Alt text](../images/architecture/scheduling_service.png)

1. The user submits Workflow to the Scheduler Service.
2. The Scheduler Service forwards user requests to the scheduler based on event scheduling.
3. The scheduler manages the behavior of Workflow, such as running jobs, stopping jobs, etc.

## Notification Server

> Notification Server: It provides notification service which support the event publishing and subscription.

Reviewer 2: ```suggestion
Notification Server: It provides Notification Service that supports the event publishing and subscription.
```

5. Metric: It provides metric meta-information services, such as model version evaluate metric, etc.
6. Artifacts: It provides artifacts meta-information services such as configuration files, jar packages etc.

### Scheduling Service

Scheduling Service: The Scheduling Service converts the workflow into an object that can be recognized by the scheduler 
and submits it to the scheduler.
The traditional scheduler can only schedule batch jobs (when upstream job finished then downstream jobs can run) 
but in the online learning scenario where we have jobs that will never finish, it does not meet the demand. 
The scheduler for Scheduler Service docking must support event-based scheduling.

The following figure shows the principle of Scheduler Service:

![Alt text](../images/architecture/scheduling_service.png)

1. The user submits Workflow to the Scheduler Service.
2. The Scheduler Service forwards user requests to the scheduler based on event scheduling.
3. The scheduler manages the behavior of Workflow, such as running jobs, stopping jobs, etc.

## Notification Server

> Notification Server: It is a service that provides event sending and event notification.

Reviewer 2: ```suggestion
Notification Server: It provides notification service which support the event publishing and subscription.
```

The figure below shows its working steps:

![Alt text](../images/architecture/notification.png)

1. A consumer listens the event which the key field equals 'Model'.
2. A producer sends the event to the Notification Server.
3. The consumer received the event, and then processes the received event according to the consumer's pre-defined logic.

## Scheduler

> Scheduler: Scheduler is responsible for workflow&job execution and support event-based scheduling.

Reviewer 2: ```suggestion
Scheduler: Scheduler is responsible for the execution of the workflow and job with the event-based scheduling mechanism.
```

The figure below shows its working steps:

![Alt text](../images/architecture/notification.png)

1. A consumer listens the event which the key field equals 'Model'.
2. A producer sends the event to the Notification Server.
3. The consumer received the event, and then processes the received event according to the consumer's pre-defined logic.

## Scheduler

Scheduler: Scheduler is responsible for executing workflow and must support event-based scheduling.

> Currently, we have implemented an [event-based scheduler](https://github.com/flink-extended/ai-flow/tree/master/lib/airflow) by combining AirFlow and Notification Server.

Reviewer 2: ```suggestion
At present, the default scheduler is the event based scheduling version of Airflow which is implemented with the Notification Service.
```
